File: workers/workerNode/app/worker.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const mqtt = require('mqtt');

// Configuraciones de conexión
const GRPC_SERVER = 'localhost:50051';
const MQTT_BROKER = 'mqtt://research.upb.edu';
const PROTO_PATH = '../proto/master.proto';

// Cargar la definición del .proto
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

// Cargar el servicio gRPC
const grpcObject = grpc.loadPackageDefinition(packageDefinition);
const masterProto = grpcObject.master;

// Crear cliente gRPC
const grpcClient = new masterProto.Master(GRPC_SERVER, grpc.credentials.createInsecure());

// Clase WorkerManager para manejar el Worker ID
class WorkerManager {
  constructor(grpcClient) {
    this.grpcClient = grpcClient;
  }

  async register(workerId) {
    return new Promise((resolve, reject) => {
      this.grpcClient.RegisterWorker({ worker_id: workerId }, (error, response) => {
        if (error) {
          console.error('Error al registrar el Worker:', error);
          reject(error);
        } else {
          console.log(`Worker registrado con ID: ${workerId}`);
          resolve(workerId);
        }
      });
    });
  }
}

// Función para enviar logs al Master
function sendLog(grpcClient, workerId, sensorId, freq, iteration) {
  return new Promise((resolve, reject) => {
    const logRequest = {
      worker_id: workerId,
      sensor_id: sensorId,
      freq: freq,
      iteration: iteration,
    };
    grpcClient.Log(logRequest, (error, response) => {
      if (error) {
        console.error('Error al enviar log:', error);
        reject(error);
      } else {
        console.log(
          `Log enviado al Master: Worker ${workerId}, Sensor ${sensorId}, Freq=${freq}, Iteration=${iteration}, Respuesta: ${response.status}`
        );
        resolve(response);
      }
    });
  });
}

// Función principal
async function main() {
  try {
    // Crear una instancia de WorkerManager y registrar el Worker
    const workerId = 'ccc'; // ID del Worker
    const workerManager = new WorkerManager(grpcClient);
    await workerManager.register(workerId);

    // Paso 2: Configurar el cliente MQTT
    const mqttClient = mqtt.connect(MQTT_BROKER, {
      port: 1883,
      keepalive: 60,
      reconnectPeriod: 1000,
      connectTimeout: 30 * 1000,
    });

    // Configurar las opciones del cliente MQTT
    mqttClient.on('connect', () => {
      console.log('Conectado al broker MQTT');

      // Suscribirse al tópico de solicitudes específico del worker
      const requestTopic = `upb/${workerId}/request`;
      mqttClient.subscribe(requestTopic, (err) => {
        if (err) {
          console.error('Error al suscribirse:', err);
        } else {
          console.log(`Worker ${workerId} suscrito al tópico: ${requestTopic}`);
        }
      });
    });

    // Manejar los mensajes MQTT recibidos
    mqttClient.on('message', async (topic, message) => {
      try {
        // Decodificar el mensaje
        const payload = JSON.parse(message.toString());
        console.log(`Mensaje recibido en ${topic}: ${JSON.stringify(payload)}`);

        // Extraer datos del mensaje
        const sensorId = payload.sensor_id;

        // Generar valores aleatorios
        const freq = Number((Math.random() * 0.9 + 0.1).toFixed(2)); // Frecuencia [0.1 - 1] con 2 decimales
        const iteration = Math.floor(Math.random() * 6 + 5); // Iteración [5 - 10] enteros

        console.log(
          `Worker ${workerId}: Tarea recibida de ${sensorId}. Freq=${freq}, Iteration=${iteration}`
        );

        // Enviar log al Master
        await sendLog(grpcClient, workerId, sensorId, freq, iteration);

        // Publicar respuesta al ESP32 en el tópico correspondiente
        const responseTopic = `upb/${workerId}/response`;
        const responseMessage = { freq, iteration };
        mqttClient.publish(responseTopic, JSON.stringify(responseMessage));
        console.log(`Respuesta publicada en ${responseTopic}: ${JSON.stringify(responseMessage)}`);
      } catch (error) {
        console.error('Error procesando mensaje:', error);
      }
    });

    // Manejar errores de MQTT
    mqttClient.on('error', (error) => {
      console.error('Error en la conexión MQTT:', error);
    });

    // Manejar desconexión
    mqttClient.on('close', () => {
      console.log('Conexión MQTT cerrada');
    });
  } catch (error) {
    console.error('Error en el Worker:', error);
  }
}

// Iniciar el worker
main();

File: workers/workerRuby/workerRuby/worker.rb
require 'grpc'
require_relative 'master_pb'
require_relative 'master_services_pb'
require 'mqtt'
require 'json'

BROKER = 'research.upb.edu'
PORT = 1883
USE_SSL = false  # Cambiar a true si se requiere SSL

# Registrar el Worker con el Master
def register_with_master(worker_id)
  stub = Master::Master::Stub.new('localhost:50051', :this_channel_is_insecure)
  request = Master::RegisterWorkerRequest.new(worker_id: worker_id)
  response = stub.register_worker(request)
  puts "Worker registrado con ID: #{worker_id}"
  return stub
end

# Enviar log al Master
def send_log(stub, worker_id, sensor_id, freq, iteration)
  request = Master::LogRequest.new(
    worker_id: worker_id,
    sensor_id: sensor_id,
    freq: freq,
    iteration: iteration
  )
  response = stub.log(request)
  puts "Log enviado: Sensor=#{sensor_id}, Freq=#{freq}, Iteraciones=#{iteration}, Status=#{response.status}"
end

# Función principal que ejecuta todo el flujo
def main
  # Paso 1: Registro con el Master
  worker_id = "bbb"
  stub = register_with_master(worker_id)
  
  # Paso 2: Configurar el cliente MQTT usando la gema 'mqtt'
  client = MQTT::Client.new(BROKER, PORT)
  client.ack_timeout = 10  # Establecer tiempo de espera para confirmaciones
  client.keep_alive = 60   # Mantener la conexión viva por 60 segundos
  client.connect

  puts "Conectado al broker MQTT"

  # Suscripción al tópico
  request_topic = "upb/#{worker_id}/request"
  client.subscribe(request_topic)
  puts "Worker #{worker_id} suscrito al tópico: #{request_topic}"

  # Configurar callback de mensaje
  client.get do |topic, message|
    begin
      puts "Mensaje recibido - Tópico: #{topic}, Payload: #{message}"
      
      msg_payload = JSON.parse(message)
      sensor_id = msg_payload['sensor_id']
      
      # Generar valores aleatorios
      freq = rand(0.1..1.0).round(2)
      iteration = rand(5..10)
      
      puts "Worker #{worker_id}: Tarea recibida de #{sensor_id}. Freq=#{freq}, Iteration=#{iteration}"
      
      send_log(stub, worker_id, sensor_id, freq, iteration)
      
      # Publicar respuesta al ESP32
      response_topic = "upb/#{worker_id}/response"
      response_message = { 'freq' => freq, 'iteration' => iteration }
      client.publish(response_topic, response_message.to_json)
      
      puts "Respuesta publicada en #{response_topic}: #{response_message}"
    rescue JSON::ParserError
      puts "Error parseando el mensaje JSON"
    rescue => e
      puts "Error procesando mensaje: #{e.message}"
    end
  end
end

main if __FILE__ == $PROGRAM_NAME
